libjpeg_addons/src/libjpg_addons.c

```diff
@@ -0,0 +1,294 @@
+	struct jpeg_compress_struct cinfo;
+	struct my_error_mgr jerr;
+	FILE *outfile;
+
+	cinfo.err = jpeg_std_error(&jerr.pub);
+	jerr.pub.error_exit = my_error_exit;
+	if (setjmp(jerr.setjmp_buffer)) {
+		jpeg_destroy_compress(&cinfo);
+		printf("jpgOpenFILE: error during processing file\n");
+		return;
+	}
+	jpeg_create_compress(&cinfo);
+
+	if ((outfile = fopen(filename, "wb")) == NULL) {
+    	fprintf(stderr, "can't open %s\n", filename);
+    	return;
+  	}
+
+  	jpeg_stdio_dest(&cinfo, outfile);
+
+	cinfo.image_width = jpg->width;
+  	cinfo.image_height = jpg->height;
+  	cinfo.input_components = jpg->bpp/8;
+  	cinfo.in_color_space = JCS_RGB;
+  
+	jpeg_set_defaults(&cinfo);
+	jpeg_set_quality(&cinfo, 100, TRUE);
+
+
+	jpeg_start_compress(&cinfo, TRUE);
+
+	unsigned int row_stride = jpg->width * 3;
+	unsigned char *row_pointer = (unsigned char *)jpg->buffer;
+	while (cinfo.next_scanline < cinfo.image_height) {
+		jpeg_write_scanlines(&cinfo, (JSAMPARRAY)&row_pointer, 1);
+		row_pointer += row_stride;
+	}
+
+	jpeg_finish_compress(&cinfo);
+  	fclose(outfile);
+
+	jpeg_destroy_compress(&cinfo);
+}
+
+#define PS2SS_GSPSMCT32		0
+#define PS2SS_GSPSMCT24		1
+#define PS2SS_GSPSMCT16		2
+
+int jpgScreenshot(const char* filename,unsigned int vramAdress, unsigned int width, unsigned int height, unsigned int psm)
+{
+	int y;
+	static uint32_t in_buffer[1024*4];  // max 1024*32bit for a line, should be ok
+	uint8_t *p_out;
+	jpgData *jpg;
+
+	jpg = malloc(sizeof(jpgData));
+	if(jpg == NULL)
+		return -1;
+
+	jpg->buffer = malloc(width * height * 3);
+	if (jpg->buffer == NULL) {
+		free(jpg);
+		return -1;
+	}
+
+	p_out = jpg->buffer;
+
+	// Check if we have a tempbuffer, if we do we use it 
+	for (y = 0; y < height; y++)
+	{
+		ps2_screenshot(in_buffer, vramAdress, 0, y, width, 1, psm);
+
+		if (psm == PS2SS_GSPSMCT16)
+		{
+			uint32_t x;
+			uint16_t* p_in  = (uint16_t*)&in_buffer;
+
+			for (x = 0; x < width; x++)
+			{
+				uint32_t r = (p_in[x] & 31) << 3;
+				uint32_t g = ((p_in[x] >> 5) & 31) << 3;
+				uint32_t b = ((p_in[x] >> 10) & 31) << 3;
+
+				p_out[x*3+0] = r;
+				p_out[x*3+1] = g;
+				p_out[x*3+2] = b;
+			}
+		}
+		else
+		{
+			if (psm == PS2SS_GSPSMCT24)
+			{
+				uint32_t x;
+				uint8_t* p_in  = (uint8_t*)&in_buffer;
+
+				for (x = 0; x < width; x++)
+				{
+					uint8_t r =  *p_in++;
+					uint8_t g =  *p_in++;
+					uint8_t b =  *p_in++;
+
+					p_out[x*3+0] = r;
+					p_out[x*3+1] = g;
+					p_out[x*3+2] = b;
+				}
+			}
+			else
+			{
+				uint8_t *p_in = (uint8_t *) &in_buffer;
+				uint32_t x;
+
+				for(x = 0; x < width; x++)
+				{
+					uint8_t r = *p_in++;
+					uint8_t g = *p_in++;
+					uint8_t b = *p_in++;
+
+					p_in++;
+
+					p_out[x*3+0] = r;
+					p_out[x*3+1] = g;
+					p_out[x*3+2] = b;
+				}
+			}
+
+			p_out+= width*3;
+		}
+	}
+
+	jpg->width = width;
+	jpg->height = height;
+	jpg->bpp = 24;
+
+	jpgFileFromJpgData(filename, 100, jpg);
```

Comment:
Quality 100? Isn't that huge, and overkill for the ps2?

Reply:
I kept same logic that was before, have in mind that this is for creating the screenshot

Reply:
For screenshots, function should be as fast as possible. With quality 100 there is no compression, so it will be the best value.

Reply:
Fast to compress or fast to store on mc/usb?

Reply:
I don't really mind to put 0-20-40-60-80-100 or put it by parameter. Here you have more info

https://libjpeg-turbo.org/About/SmartScale-Lossy

Reply:
So compression times can be really slow (slower than storage speed). I guess 100 is a good value then. Thanks.